### PR TITLE
Synchronize changes hmat-devel -> main

### PR DESCRIPTION
**Pull request by michalbiesek**
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/500)

### CODE REVIEW

**Review by PatKamin — COMMENTED** [review]
> Reviewed 88 of 88 files at r1.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/500)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @michalbiesek)

**Review by PatKamin — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/500#-:-MSD3qsg2yAubQ8DdWq9:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @michalbiesek)

**Review by pbalcer — COMMENTED** [review]
Overall, the API lgtm (I looked over the code, but I didn't do a full review). Couple of things though:
- How will this new API behave if there's no HMAT in the platform? This is not described.
- I think you need to better explain things in the man page on how you detect things like capacity/latency (related to the previous point).
- I'd like to see a simple example demonstrating how this can be utilized (maybe as a next step, to avoid making this PR even bigger).

**Comment by michalbiesek** [comment]
> Overall, the API lgtm (I looked over the code, but I didn't do a full review). Couple of things though:
> 
> * How will this new API behave if there's no HMAT in the platform? This is not described.
> * I think you need to better explain things in the man page on how you detect things like capacity/latency (related to the previous point).

Thanks, I try to address these concerns in #507, by adding requirements description of new memory kinds. 
Be aware that for capacity memory characteristics I don't depend on HMAT at all.

> * I'd like to see a simple example demonstrating how this can be utilized (maybe as a next step, to avoid making this PR even bigger).

I plan to provide an example usage of this new API as a post on the blog.

File: README [hunk 1]
```diff
@@ -302,6 +327,15 @@ threads to CPUs associated with NUMA node 0.  With this configuration
 standard allocations will be fetched across the QPI bus, and high
 bandwidth allocations will be local to the process CPU.
 
+## Identification of High Bandwidth Memory
+High bandwidth memory are identified using hwloc if available. The default
```
Comment (mjsmolarek) [inline README:331]: Please correct to (mind the comma too):
High bandwidth memory is identified using hwaloc, if available

File: configure.ac [hunk 2]
```diff
@@ -269,4 +294,13 @@ AC_MSG_RESULT([
                To enable this feature please install libdaxctl-devel v66 or later
                and run configure again])
 fi
+if test $hwloc_hmat = "yes" ; then
+AC_MSG_RESULT([
+               Utilize information about memory performance characteristics is enabled])
```
Comment (mjsmolarek) [inline configure.ac:299]: Please correct: Utilization of information

File: configure.ac [hunk 3]
```diff
@@ -269,4 +294,13 @@ AC_MSG_RESULT([
...
+else
+AC_MSG_RESULT([
+               Utilize information about memory performance characteristics is disabled
```
Comment (mjsmolarek) [inline configure.ac:302]: Please correct: Utilization of information

File: man/memkind.3 [hunk 4]
```diff
@@ -552,7 +552,7 @@ is available or an error code from the
 section if it is not.
 .PP
 .BR memkind_check_dax_path ()
-retruns zero if file-backed kind memory in the specified directory path
+returns zero if file-backed kind memory in the specified directory path
```
Comment (mjsmolarek) [inline man/memkind.3:555]: Missing a verb here...
Returns zero if file-backed kind memory is (?) in the specified directory path

File: man/memkind.3 [hunk 5]
```diff
@@ -688,6 +688,57 @@ The available kinds of memory:
 .B MEMKIND_DEFAULT
 Default allocation using standard memory and default page size.
 .TP
+.B MEMKIND_HIGHEST_CAPACITY
+Allocate from a NUMA node(s) that has the highest capacity among all nodes in the system.
+.TP
+.B MEMKIND_HIGHEST_CAPACITY_PREFERRED
+Same as
```
Comment (mjsmolarek) [inline man/memkind.3:695]: "Same as above" or "Same as MEMKIND_HIGHEST_CAPACITY"

File: README (hunk 1 above)
Comment (michalbiesek) [reply README:331]: Done in #507 ( after merge #507 it will be available in this PR)

File: configure.ac (hunk 2 above)
Comment (michalbiesek) [reply configure.ac:299]: Done in #507 ( after merge #507 it will be available in this PR)

File: configure.ac (hunk 3 above)
Comment (michalbiesek) [reply configure.ac:302]: Done in #507 ( after merge #507 it will be available in this PR)

File: man/memkind.3 (hunk 4 above)
Comment (michalbiesek) [reply man/memkind.3:555]: Done in #507 ( after merge #507 it will be available in this PR)

File: man/memkind.3 (hunk 5 above)
Comment (michalbiesek) [reply man/memkind.3:695]: It is "Same as MEMKIND_HIGHEST_CAPACITY" see https://github.com/memkind/memkind/blob/46e6b883665a9eda651f4ee775c214282dbe6651/man/memkind.3#L694-L696

File: man/memkind.3 [hunk 6]
```diff
@@ -986,6 +1044,11 @@ for parsing, so the syntax described in the
 .BR numa (3)
 man page for this routine applies: e.g. 1-3,5 is a valid setting.
 .TP
+.B MEMKIND_HBW_THRESHOLD
+This environment variable is bandwidth in MB/s that is the threshold for
+identifying high bandwidth memory. The default threshold is 204800 (200 GB/s),
+which is used if this environment is not set.
```
Comment (mjsmolarek) [inline man/memkind.3:1050]: "if this variable is not set" or maybe use the abbreviation "if this EV is not set"

File: utils/docker/README.md [hunk 7]
```diff
@@ -21,22 +21,34 @@ to build, test and optionally measure test coverage of any pull request to memki
 For fully testing MEMKIND_DAX_KMEM, ndctl library version tag must be passed as parameter,
 see https://github.com/pmem/ndctl/tags.
 
+* **ENABLE_HWLOC** - install hwloc library.
+For fully testing MEMKIND_HBW*, MEMKIND_HIGHEST_BANDWIDTH, MEMKIND_LOWEST_LATENCY, MEMKIND_*LOCAL hwloc library version must be installed.
```
Comment (mjsmolarek) [inline utils/docker/README.md:25]: Please change to: To fully test

"hwloc library version must be installed" - a specific version or any version of the library?

Comment (michalbiesek) [reply utils/docker/README.md:25]: Done in #507 ( after merge #507 it will be available in this PR)

File: man/memkind.3 (hunk 6 above)
Comment (michalbiesek) [reply man/memkind.3:1050]: Done in #507 ( after merge #507 it will be available in this PR)

**Pull request merged**